geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/KryoLazyDensityUtils.scala

```diff
@@ -18,48 +18,46 @@ import org.locationtech.geomesa.utils.geotools.GridSnap
 import org.locationtech.geomesa.utils.interop.SimpleFeatureTypes
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 import org.opengis.filter.expression.Expression
+import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
 
 import scala.collection.mutable
 
-trait KryoLazyDensityUtils {
+class KryoLazyDensityUtils(options: Map[String, String], sft: SimpleFeatureType) {
 
   import KryoLazyDensityUtils._
-  var geomIndex: Int = -1
+
+  val geomIndex: Int = Option(sft.getGeomIndex).getOrElse(-1)
   // we snap each point into a pixel and aggregate based on that
-  var gridSnap: GridSnap = null
-  var writeGeom: (SimpleFeature, DensityResult) => Unit = null
-
-  def initialize(options: Map[String, String], sft: SimpleFeatureType): DensityResult = {
-    import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
-    geomIndex = sft.getGeomIndex
-    gridSnap = {
-      val bounds = options(ENVELOPE_OPT).split(",").map(_.toDouble)
-      val envelope = new Envelope(bounds(0), bounds(1), bounds(2), bounds(3))
-      val Array(width, height) = options(GRID_OPT).split(",").map(_.toInt)
-      new GridSnap(envelope, width, height)
+  val gridSnap = {
+    val bounds = options(ENVELOPE_OPT).split(",").map(_.toDouble)
+    val envelope = new Envelope(bounds(0), bounds(1), bounds(2), bounds(3))
+    val Array(width, height) = options(GRID_OPT).split(",").map(_.toInt)
+    new GridSnap(envelope, width, height)
+  }
+  // function to get the weight from the feature - defaults to 1.0 unless an attribute/exp is specified
+  val weightIndex = options.get(WEIGHT_OPT).map(sft.indexOf).getOrElse(-2)
+
+  def newDensityResult: DensityResult = mutable.Map.empty[(Int, Int), Double]
+
+  def weightFn(sf: SimpleFeature): Double = {
```

Comment:
this should stay as a val so we don't e.g. evaluate the expression on line 46 every time